Fixed #33605 -- Fixed migration crash when altering RegexValidator to pre-compiled regular expression.

brianhelba opened the pull request:
https://code.djangoproject.com/ticket/33605

felixxm commented on django/db/migrations/utils.py:
```diff
@@ -10,7 +10,11 @@ def __init__(self, obj):
 
     def __eq__(self, other):
-        return self.pattern == other.pattern and self.flags == other.flags
+        return (
+            isinstance(other, RegexObject)
+            and self.pattern == other.pattern
+            and self.flags == other.flags
+        )
```
This should return `NotImplemented` when a type of `other` doesn't match:
```suggestion
        if not isinstance(other, RegexObject):
            return NotImplemented
        return self.pattern == other.pattern and self.flags == other.flags
```

brianhelba replied:
If `other` is a string (typically representing a regex), then it's reasonable to return `False` from the comparison, which will trigger the migration engine to generate a new migration.

In general, I see nothing wrong with failed `__eq__` comparations returning `False`. This is how every other equivalence comparison in Python works.

More importantly, raising a different exception (`NotImplemented` instead of `AttributeError`) doesn't solve the practical bug, which is that it should be possible to change a `RegexValidator` from using a `str` to a `re.compile` (or vice versa) and have that successfully generate a migration.

felixxm replied:
We should **return** `NotImplemented` not raise a different exception, see 54ea290e5bbd19d87bd8dba807738eeeaf01a362 and ticket-30651.

brianhelba replied:
Ah, I understand. Sorry for my confusion.

This is fixed.

felixxm reviewed (commented):
@brianhelba Thanks for the patch :+1: Please add a regression test.

brianhelba commented:
I can certainly add a new test for this (such will fail without the change, and pass with it).

mohamed-rasal commented:
guys

brianhelba commented:
@felixxm I've updated the implementation and added a test. O've confirmed that the test fails without the change and succeeds with it.

This is ready for re-review.

felixxm reviewed (approved):
@brianhelba Thanks :+1:
